### PR TITLE
backend-app-api: filter internal errors

### LINKED ISSUE #22982: API calls return postgresql query string in browser
### 📜 Description

The backstage API returns the query string when we pass "%00" as filter. Is this a known issue?

**- api/catalog/entity-facets?facet=%00**
`{
"error": {
"name": "error",
"message": "select \"search\".\"original_value\" as \"value\", count(*) as \"count\" from \"search\" where \"search\".\"key\" = $1 and \"search\".\"original_value\" is not null group by \"search\".\"original_value\" - invalid byte sequence for encoding \"UTF8\": 0x00",
"length": 146,
"severity": "ERROR",
"code": "22021",
"where": "unnamed portal parameter $1",
"file": "mbutils.c",
"line": "1669",
"routine": "report_invalid_encoding"
},
"request": {
"method": "GET",
"url": "[/entity-facets?facet=%00](******/entity-facets?facet=%00)"
},
"response": {
"statusCode": 500
}
}`

**- /api/catalog/entities?filter=kind=%00&filter=kind=group**
`{
"error": {
"name": "error",
"message": "select \"final_entities\".* from \"final_entities\" where \"final_entities\".\"final_entity\" is not null and (((((\"final_entities\".\"entity_id\" in (select \"search\".\"entity_id\" from \"search\" where \"key\" = $1 and (\"value\" = $2)))))) or ((((\"final_entities\".\"entity_id\" in (select \"search\".\"entity_id\" from \"search\" where \"key\" = $3 and (\"value\" = $4))))))) order by \"final_entities\".\"entity_id\" asc - invalid byte sequence for encoding \"UTF8\": 0x00",
"length": 146,
"severity": "ERROR",
"code": "22021",
"where": "unnamed portal parameter $2",
"file": "mbutils.c",
"line": "1669",
"routine": "report_invalid_encoding"
},
"request": {
"method": "GET",
"url": "[/entities?filter=kind=%00&filter=kind=group](*****entities?filter=kind=%00&filter=kind=group)"
},
"response": {
"statusCode": 500
}
}`

### 👍 Expected behavior

It should just return not found. The error message should not contain sql statements.

### 👎 Actual Behavior with Screenshots

See description

### 👟 Reproduction steps

See description

### 📃 Provide the context for the Bug.

_No response_

### 🖥️ Your Environment

_No response_

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

None

### PR DESCRIPTION
**Pull request by Rugvip**
## Hey, I just made a Pull Request!

This adds a new error filtering capability to the default error handling. The purpose is to filter out errors that we never want to return in a response, and instead log it along with a correlation ID.

Fixes #22982

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/backend-app-api/src/http/applyInternalErrorFilter.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ * limitations under the License.
+ */
+
+import { LoggerService } from '@backstage/backend-plugin-api';
+import { assertError } from '@backstage/errors';
+import { randomBytes } from 'crypto';
+
+function handleBadError(error: Error, logger: LoggerService) {
+  const logId = randomBytes(10).toString('hex');
+  logger.error(
```
Comment (freben) [inline packages/backend-app-api/src/http/applyInternalErrorFilter.ts:23]: would make sense to have the log id as meta info so it ends up in json formatted logs as a dedicated field right

**Review by freben — APPROVED** [review]
just one optional question

File: packages/backend-app-api/src/http/applyInternalErrorFilter.ts (hunk 1 above)
Comment (Rugvip) [reply packages/backend-app-api/src/http/applyInternalErrorFilter.ts:23]: Yeah would be nice, but we can't have both that and the error fields being expanded. Might not be what we want though tbh, as in errors being expanded. I'll have a look at what a middle ground could be

Comment (Rugvip) [reply packages/backend-app-api/src/http/applyInternalErrorFilter.ts]: Alright got it in there with a child logger

**Review by vinzscam — APPROVED** [review]
nice

**Pull request merged**